gtsam/linear/SparseEigenSolver.cpp

```diff
@@ -0,0 +1,231 @@
+/* ----------------------------------------------------------------------------
+
+ * GTSAM Copyright 2010, Georgia Tech Research Corporation,
+ * Atlanta, Georgia 30332-0415
+ * All Rights Reserved
+ * Authors: Frank Dellaert, et al. (see THANKS for the full author list)
+
+ * See LICENSE for the license information
+
+ * -------------------------------------------------------------------------- */
+
+/**
+ * @file SparseEigenSolver.cpp
+ *
+ * @brief Eigen SparseSolver based linear solver backend for GTSAM
+ *
+ * @date Aug 2019
+ * @author Mandy Xie
+ * @author Fan Jiang
+ * @author Frank Dellaert
+ */
+
+#include <gtsam/base/timing.h>
+#include <gtsam/linear/SparseEigenSolver.h>
+#include <vector>
+
+#include <boost/shared_ptr.hpp>
+
+using namespace std;
+
+namespace gtsam {
+
+  using SpMat = Eigen::SparseMatrix<double, Eigen::ColMajor, Eigen::Index>;
+
+  Eigen::SparseMatrix<double>
+  SparseEigenSolver::sparseJacobianEigen(
+      const GaussianFactorGraph &gfg,
+      const Ordering &ordering) {
+    // First find dimensions of each variable
+    std::map<Key, size_t> dims;
+    for (const boost::shared_ptr<GaussianFactor> &factor : gfg) {
+      if (!static_cast<bool>(factor))
+        continue;
+
+      for (auto it = factor->begin(); it != factor->end(); ++it) {
+        dims[*it] = factor->getDim(it);
+      }
+    }
+
+    // Compute first scalar column of each variable
+    size_t currentColIndex = 0;
+    std::map<Key, size_t> columnIndices;
+    for (const auto key : ordering) {
+      columnIndices[key] = currentColIndex;
+      currentColIndex += dims[key];
+    }
+
+    // Iterate over all factors, adding sparse scalar entries
+    vector<Eigen::Triplet<double>> entries;
+    entries.reserve(60 * gfg.size());
```

Comment:
Is 60 just an arbitrary "reasonably liberal" # of nonzeros per factor?

Reply:
Yes, just to prevent reallocation (to an extent) since it's pretty much impossible to estimate the actual size